### DISCUSSION TITLE
Reverse Z buffer

### DISCUSSION THREAD

**Discussion by voxelias** [gfx-rs/wgpu#1497] [Q&A] [answered]
What I want to achieve is a possibility of rendering of a huge objects (lets say an Island) from a large distance.  I have a character on that island and I want to move camera far away from it to observe all the island. In other words I want to get rid of clipping effect of the projection matrix.

I've read that one possibility is to use reverse Z buffer. And here is a [nice article](https://nlguillemot.wordpress.com/2016/12/07/reversed-z-in-opengl/) about it.

But I am a bit lost, what should I do to achieve that in WGPU context and terms?
Any help and guidance would be very appreciated :)

**Comment by cwfitzgerald** [comment, marked as answer]
In terms of wgpu, all you need it to use a floating point depth buffer (Depth32Float). Then on the math end, you want your projection matrix to invert Z. A lot of graphics focused math libs can generate these (see for example, glam's [`Mat4::perspective_infinite_reverse_lh`](https://docs.rs/glam/0.16.0/glam/f32/struct.Mat4.html#method.perspective_infinite_reverse_lh)). What are you currently using to generate your projection matrix?

**Reply by kvark** [reply]
And also use `GreaterEqual` comparison in the depth-stencil state of the pipeline, of course :)

**Reply by voxelias** [reply]
Thank you for the response! I already use `Depth32Float`:

```
        let texture = wgpu::TextureDescriptor {
            label: Some("Depth Buffer"),
            size: buffer_extent,
            mip_level_count: 1,
            sample_count: 1,
            dimension: wgpu::TextureDimension::D2,
            format: wgpu::TextureFormat::Depth32Float,
            usage: wgpu::TextureUsage::SAMPLED
                | wgpu::TextureUsage::COPY_DST
                | wgpu::TextureUsage::RENDER_ATTACHMENT,
        };
```

Is that correct?

To generate my projection matrix at this moment I use `cgmath` but that is not an issue. I can change the lib or implement the matrix building function.

**Reply by kvark** [reply]
this looks fine

**Reply by voxelias** [reply]
Thank you, I will try different projection matrix!
